### ISSUE TITLE
RFC 140 migration plan

### ISSUE THREAD

**Issue by infinisil** [NixOS/nixpkgs#258650]
Here's my current plan to implement [part 2 of RFC 140](https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md#pr-2-automated-migration):

0. Implement a strategy to make it easy to increase the strictness of the `pkgs/by-name` CI check:
   - https://github.com/NixOS/nixpkgs/issues/256788
   - https://github.com/NixOS/nixpkgs/pull/272395
2. Require packages that use `pkgs/by-name` to not be defined with an `{ }` argument in `all-packages.nix`:
   - https://github.com/NixOS/nixpkgs/pull/256792
   - https://github.com/NixOS/nixpkgs/pull/272395
   - https://github.com/NixOS/nixpkgs/pull/274591
4. Require new packages using `callPackage` to use `pkgs/by-name`
   - https://github.com/NixOS/nixpkgs/pull/278805
   - https://github.com/NixOS/nixpkgs/pull/275539
6. Try to migrate all packages that don't cause failures or output path changes
   - https://github.com/NixOS/nixpkgs/pull/354531
   - still left to do:
     - defined in individual files
     - with override (`callPackage ./foo.nix { foo = bar; }`)

Ping the Nixpkgs Architecture Team, @roberth @Ericson2314 @tomberek @aakropotkin @DavHau @phaer

**Comment by SuperSandro2000** [comment]
What if I don't want my packages to be trapped in this experimental feature? It doesn't solve any problem for me personally, so it can only create new ones like breaking [this fancy prepare-commit-msg hook](https://gist.github.com/SuperSandro2000/77439b7eb5f7a0da6967524190a4ce8f) I am using for a year or it will likely break brittle update scripts ([like the one for ArchiSteamFarm](https://github.com/NixOS/nixpkgs/commit/c781bd7a53cf38594292e310a22e14a565d8a746) which are already pretty hard to test and I honestly don't know if it would work for r-rynatm even.

**Comment by phaer** [comment]
> What if I don't want my packages to be trapped in this experimental feature? 

Than you would have had about 7 months time between the initial proposal and the end of FCP to add your concerns to the [extensive discussion](https://github.com/NixOS/rfcs/pull/140). You did so just now, so I would like to kindly ask you to keep this kind of the debate out of this thread which explicitly is about the implementation only. :)

**Comment by amjoseph-nixpkgs** [comment]
> What if I don't want my packages

We don't have package owners.  Everybody is allowed to submit PRs against any package.

> to be trapped in this experimental feature?

This isn't an experimental feature; it's the future and it's long overdue.  I have plenty of automation that gets broken constantly because there is no reliable way to map from attrnames to files.  Overlays get silently disabled when I rebase nixpkgs because there's no restrictions on what code can see which files -- I can't count the number of times my override-patches have been silently dropped because somebody rewired `all-packages.nix` and found a new way to reference the package's definition without going through the attrname I'd overridden.  It's madness.

**Comment by SuperSandro2000** [comment]
> I have plenty of automation that gets broken constantly because there is no reliable way to map from attrnames to files.

There will be now for packages that are in this but there will be still plenty of interesting packages left, especially the more complicated ones.

> Overlays get silently disabled when I rebase nixpkgs because there's no restrictions on what code can see which files -- I can't count the number of times my override-patches have been silently dropped because somebody rewired `all-packages.nix` and found a new way to reference the package's definition without going through the attrname I'd overridden. It's madness.

Except for versioned attrs people should create aliases which should make this manageable.

I also have a [good forest of extra commits](https://github.com/NixOS/nixpkgs/compare/nixos-unstable...SuperSandro2000:nixpkgs:nixos-unstable) and overlays and I mostly touch them when things changed in the exact same files I was changing. Changing some things with overrides is a bit more complicated than directly touching the code especially if you plan to upstream it anyway.

With this change still everything sees everything. I don't see how that definitely solves your problems.

> We don't have package owners. Everybody is allowed to submit PRs against any package.

If I tell that some people about some of their beloved packages then they would walk up the next wall.

Also this contradicts what you wrote below. If anyone can change anything at will then sure things break for others break left, right and center. Be it your overlays or my commits.

**Comment by ehmry** [comment]
What happens to `libsForQt5.callPackage ././. { }` calls in `all-packages.nix`?

**Comment by infinisil** [comment]
@ehmry Nothing, see https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md#drawbacks or https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#limitations

**Comment by infinisil** [comment]
This PR makes progress on points 0. and 1.: https://github.com/NixOS/nixpkgs/pull/272395

**Comment by blaggacao** [comment]
@infinisil Maybe there's a way to reconcile the concerns about breaking update scripts. A regex fishing for `../..` patterns could with some probablility identify update scripts that break out of their unit folder-boundary and would thus break. These could then be put in another shortlist and scheduled for manual migration. Would that be a viable way out for the migration addressing @SuperSandro2000 's concerns?

**Comment by infinisil** [comment]
@blaggacao I think with the gradual transition enabled by approaches like https://github.com/NixOS/nixpkgs/pull/272395 that's totally a possibility. Once I get to running the automation, I'll do some quick grepping for `../..`, `update.sh` and `updateScript` to hopefully catch most. Those would then just be excluded from the automated migration and have to be fixed manually.

**Comment by infinisil** [comment]
Two new PRs:
- https://github.com/NixOS/nixpkgs/pull/274591 is ready to be reviewed/merged. It enables https://github.com/NixOS/nixpkgs/pull/272395 in CI
- https://github.com/NixOS/nixpkgs/pull/275539 enforces new packages to use `pkgs/by-name`

**Comment by katexochen** [comment]
> 3. Try to migrate all packages that don't cause failures or output path changes
https://github.com/NixOS/nixpkgs/pull/211832

Is the automatic migration still planned? Is this the next step, or do we need anything else?

**Comment by infinisil** [comment]
@katexochen Indeed I'm still planning to do that, I'll try to get it done as soon as possible 😃

**Comment by Atemu** [comment]
Hi, what's the progress on this?

**Comment by infinisil** [comment]
@Atemu In recent weeks we moved the tooling to a separate repo and got more people involved in contributions, so development speed is going faster now: https://github.com/NixOS/nixpkgs-check-by-name

Still not done, but work in on-going :)

**Comment by infinisil** [comment]
I just wrote down this thorough description of the current state and plan regarding the implementation of the automated migration: https://github.com/NixOS/nixpkgs-check-by-name/issues/56

tl;dr: It's quite tricky but doable!
